Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm

```diff
@@ -107,9 +110,75 @@
     // FIXME: rdar://118476774 - Support session rules
 }
 
-void WebExtensionAPIDeclarativeNetRequest::getMatchedRules(NSDictionary *filter, Ref<WebExtensionCallbackHandler>&&, NSString **outExceptionString)
+static bool extensionHasPermission(WebExtensionContextProxy& extensionContext, NSString *permission)
 {
-    // FIXME: rdar://118940129 - Support getMatchedRules
+    // FIXME: https://webkit.org/b/259914 This should be a hasPermission: call to extensionContext() and updated with actually granted permissions from the UI process.
+    auto *permissions = objectForKey<NSArray>(extensionContext.manifest(), @"permissions", true, NSString.class);
+    return [permissions containsObject:permission];
+}
+
+static NSDictionary *toWebAPI(const Vector<WebExtensionMatchedRuleParameters>& matchedRules)
+{
+    NSMutableArray *matchedRuleArray = [NSMutableArray array];
+    for (auto& matchedRule : matchedRules) {
+        auto urlString = matchedRule.url.string();
+        auto tabIdentifier = matchedRule.tabIdentifier;
+        auto timeStamp = matchedRule.timeStamp;
+
+        [matchedRuleArray addObject:@{
+            @"request": @{ @"url": urlString },
+            @"timeStamp": @(timeStamp),
```

Comment:
`@(floor(timeStamp.secondsSinceEpoch().milliseconds()))`

Reply:
While I'm here `WebExtensionContextProxy::dispatchWebNavigationEvent` should do `floor()` too!